pkg/storage/chunk/prometheus_chunk.go

```diff
@@ -230,8 +230,11 @@ func (p *prometheusChunkIterator) Timestamp() int64 {
 	return p.it.AtT()
 }
 
-func (p *prometheusChunkIterator) Batch(size int, valueType chunkenc.ValueType) Batch {
-	var batch Batch
+func (p *prometheusChunkIterator) Batch(size int, valueType chunkenc.ValueType, b *Batch) *Batch {
+	batch := b
+	if batch == nil {
+		batch = &Batch{}
+	}
 	batch.ValueType = valueType
```

Comment:
Here we're losing the information about the type stored in the provided `b` Batch. So it cat happen that `b` is a FloatHistogram type with FloatHistograms stored in the pointers and we're going to try to overwrite with integer Histogram.

Reply:
The above scenario should very rarely, so I think the simplest way to deal with it is to not reuse `b` if the types don't match.